File: kuji-lab/types/room.ts
// Mirror of kuji-server/models/room.py Pydantic models

export interface MemberInfo {
  user_id: string
  name: string
  avatar?: string | null
  joined_at: string
  draws_total: number
  connected: boolean
}

export interface DrawResult {
  seq: number
  user_id: string
  user_name: string
  grade: string
  prize_name: string
  prize_image?: string | null
  tickets_left: number
  drawn_at: string
}

export interface GradeTally {
  grade: string
  count: number
}

export interface RoomSnapshot {
  code: string
  slug: string
  title: string
  price_yen?: number | null
  status: 'waiting' | 'active' | 'finished'
  draws_per_turn: number
  tickets_left: number
  total_tickets: number
  pool: Record<string, number>
  members: MemberInfo[]
  queue: string[]
  results: DrawResult[]
  host_id: string
  you: string
}

// ── WebSocket events (server → client) ───────────────────────────────────────

export type RoomEvent =
  | { type: 'room_state' } & RoomSnapshot
  | { type: 'member_joined'; member: MemberInfo }
  | { type: 'member_left'; user_id: string; name?: string; reason?: string }
  | { type: 'game_started'; queue: string[] }
  | { type: 'your_turn'; draws_allowed: number; timeout_at: string }
  | { type: 'draw_result'; result: DrawResult; pool: Record<string, number> }
  | { type: 'turn_advanced'; user_id: string; reason: string; next_user_id: string | null }
  | { type: 'turn_skipped'; user_id: string; reason: string; next_user_id: string | null }
  | { type: 'pool_update'; pool: Record<string, number> }
  | { type: 'game_finished'; results_summary: GradeTally[] }
  | { type: 'host_transferred'; new_host_id: string }
  | { type: 'settings_changed'; draws_per_turn: number }
  | { type: 'kicked' }
  | { type: 'error'; code: string; message: string }
  | { type: 'pong'; ts: number }

File: kuji-server/models/room.py
from __future__ import annotations
from typing import Literal, Optional
from pydantic import BaseModel


class MemberInfo(BaseModel):
    user_id: str
    name: str
    avatar: Optional[str] = None
    joined_at: str
    draws_total: int = 0
    connected: bool = True


class DrawResult(BaseModel):
    seq: int
    user_id: str
    user_name: str
    grade: str
    prize_name: str
    prize_image: Optional[str] = None
    tickets_left: int
    drawn_at: str


class GradeTally(BaseModel):
    grade: str
    count: int


class RoomSnapshot(BaseModel):
    code: str
    slug: str
    title: str
    price_yen: Optional[int] = None
    status: Literal["waiting", "active", "finished"]
    draws_per_turn: int
    tickets_left: int
    total_tickets: int
    pool: dict[str, int]
    members: list[MemberInfo]
    queue: list[str]
    results: list[DrawResult]
    host_id: str
    you: str


class CreateRoomRequest(BaseModel):
    slug: str
    title: str
    price_yen: Optional[int] = None
    prizes: list[dict]   # raw Prize dicts from frontend
    draws_per_turn: int = 1
    # Identity is derived from the signed auth token, not these fields; they are
    # accepted for backward compatibility but ignored by the server.
    user_id: Optional[str] = None
    user_name: Optional[str] = None
    user_avatar: Optional[str] = None


class CreateRoomResponse(BaseModel):
    code: str
    snapshot: RoomSnapshot
